src/Simplex/Messaging/Server/MsgStore/Journal.hs

```diff
@@ -0,0 +1,556 @@
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Simplex.Messaging.Server.MsgStore.Journal
+  ( JournalMsgStore (msgQueues, random),
+    JournalMsgQueue,
+    JournalStoreConfig (..),
+    getQueueMessages,
+    -- below are exported for tests
+    MsgQueueState (..),
+    JournalState (..),
+    msgQueueDirectory,
+    readWriteQueueState,
+    newMsgQueueState,
+    newJournalId,
+    logQueueState,
+    queueLogFileName,
+    logFileExt,
+  )
+where
+
+import Control.Concurrent.STM
+import qualified Control.Exception as E
+import Control.Logger.Simple
+import Control.Monad
+import Control.Monad.Trans.Except
+import qualified Data.Attoparsec.ByteString.Char8 as A
+import Data.ByteString.Char8 (ByteString)
+import qualified Data.ByteString.Char8 as B
+import qualified Data.ByteString.Lazy.Char8 as LB
+import Data.Functor (($>))
+import Data.Maybe (catMaybes, fromMaybe)
+import qualified Data.Text as T
+import Data.Time.Clock (getCurrentTime)
+import Data.Time.Format.ISO8601 (iso8601Show)
+import GHC.IO (catchAny)
+import Simplex.Messaging.Agent.Client (getMapLock, withLockMap)
+import Simplex.Messaging.Agent.Lock
+import Simplex.Messaging.Encoding.String
+import Simplex.Messaging.Protocol (ErrorType (..), Message (..), RecipientId)
+import Simplex.Messaging.Server.MsgStore.Types
+import Simplex.Messaging.TMap (TMap)
+import qualified Simplex.Messaging.TMap as TM
+import Simplex.Messaging.Util (ifM, tshow, ($>>=))
+import System.Directory
+import System.FilePath ((</>))
+import System.IO (BufferMode (..), Handle, IOMode (..), SeekMode (..), stdout)
+import qualified System.IO as IO
+import System.Random (StdGen, genByteString, newStdGen)
+
+data JournalMsgStore = JournalMsgStore
+  { config :: JournalStoreConfig,
+    random :: TVar StdGen,
+    queueLocks :: TMap RecipientId Lock,
+    msgQueues :: TMap RecipientId JournalMsgQueue
+  }
+
+data JournalStoreConfig = JournalStoreConfig
+  { storePath :: FilePath,
+    pathParts :: Int,
+    quota :: Int,
+    -- Max number of messages per journal file - ignored in STM store.
+    -- When this limit is reached, the file will be changed.
+    -- This number should be set bigger than queue quota.
+    maxMsgCount :: Int,
+    maxStateLines :: Int
+  }
+
+data JournalMsgQueue = JournalMsgQueue
+  { config :: JournalStoreConfig,
+    queueDirectory :: FilePath,
+    queueLock :: Lock,
+    state :: TVar MsgQueueState,
+    -- Last message and length incl. newline
+    -- Nothing - unknown, Just Nothing - empty queue.
+    -- This optimization  prevents reading each message at least twice,
+    -- or reading it after it was just written.
+    tipMsg :: TVar (Maybe (Maybe (Message, Int))),
+    handles :: TVar (Maybe MsgQueueHandles),
+    random :: TVar StdGen
+  }
+
+data MsgQueueState = MsgQueueState
+  { writeState :: JournalState,
+    readState :: JournalState,
```

Comment:
does JournalState have the same meaning in read or write? e.g. "Read state" in both cases, for write state - once write journal becomes read journal? or is it write/read position for write/read state - so meaning is different? maybe better to have different types to not confuse these?

Reply:
will parameterize